src/main/java/core/basesyntax/service/ShopServiceImpl.java

```diff
@@ -0,0 +1,30 @@
+package core.basesyntax.service;
+
+import core.basesyntax.dao.FruitOperationDao;
+import core.basesyntax.model.FruitOperation;
+import core.basesyntax.strategy.OperationStrategy;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    private FruitOperationDao fruitOperationDao;
+    private OperationStrategy strategy;
+
+    public ShopServiceImpl(FruitOperationDao fruitOperationDao, OperationStrategy strategy) {
+        this.fruitOperationDao = fruitOperationDao;
+        this.strategy = strategy;
+    }
+
+    @Override
+    public void changeQuantityStore(List<FruitOperation> fruits) {
+        for (FruitOperation fruit : fruits) {
+            if (fruit.getOperation() == FruitOperation.Operation.BALANCE) {
+                fruitOperationDao.add(fruit);
+            }
+            FruitOperation currentFruit = fruitOperationDao.get(fruit.getFruit());
+            int newValueFrom = strategy.get(fruit.getOperation())
+                    .getQuantity(currentFruit.getQuantity(), fruit.getQuantity());
+            currentFruit.setQuantity(newValueFrom);
```

Comment:
Before setting the new quantity with `currentFruit.setQuantity(newValueFrom)`, you should check if the new value is valid. For instance, if the operation is 'PURCHASE', the quantity should not be negative. Throw a `RuntimeException` if the resulting quantity is invalid, as per the checklist.